core/arch/arm/plat-synquacer/timer_fiq.c

```diff
@@ -0,0 +1,74 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2018, Linaro Limited
+ */
+
+#include <arm.h>
+#include <console.h>
+#include <drivers/gic.h>
+#include <io.h>
+#include <kernel/panic.h>
+#include <kernel/misc.h>
+#include <rng_pta.h>
+
+static inline uint64_t read_cntps_ctl_el1(void)
+{
+	uint64_t v;
+
+	asm volatile ("mrs %0, cntps_ctl_el1" : "=r" (v));
+
+	return v;
+}
+
+static inline void write_cntps_ctl_el1(uint64_t v)
+{
+	asm volatile ("msr cntps_ctl_el1, %0" : : "r" (v));
+}
+
+static inline uint64_t read_cntps_cval_el1(void)
+{
+	uint64_t v;
+
+	asm volatile ("mrs %0, cntps_cval_el1" : "=r" (v));
+
+	return v;
+}
+
+static inline void write_cntps_cval_el1(uint64_t v)
+{
+	asm volatile ("msr cntps_cval_el1, %0" : : "r" (v));
+}
+
+void generic_timer_start(void)
+{
+	uint64_t cval;
+	uint32_t ctl = 1;
+
+	/* The timer will fire every 2 ms */
+	cval = read_cntpct() + (read_cntfrq() / 500);
+	write_cntps_cval_el1(cval);
+
+	/* Enable the secure physical timer */
+	write_cntps_ctl_el1(ctl);
+}
+
+void generic_timer_stop(void)
+{
+	/* Disable the timer */
+	write_cntps_ctl_el1(0);
+}
+
+void generic_timer_handler(void)
+{
+	/* Ensure that the timer did assert the interrupt */
+	assert((read_cntps_ctl_el1() >> 2));
+
+	/*
+	 * Disable the timer and reprogram it. The barriers ensure that there is
+	 * no reordering of instructions around the reprogramming code.
+	 */
+	isb();
```

Comment:
That's a curious way of using `isb()`. Which reordering are you trying to protect against?

Reply:
Reordering of instructions which accesses cntps_ctl_el1 register. Please suggest if there is another better way of handling this.

Reply:
Who's doing the reordering? Since these are inline assembly instructions I don't think the compiler will reorder anything, if it does we've implemented the inline assembly wrong.

Reply:
Ok got your point. Will remove this `isb()`.